[vcpkg] Meson osx sysroot

daschuer opened the pull request:
This allows to use VCPKG_OSX_SYSROOT with meson ports. This is required to crosscompile formatconfig for arm64 on osx.  Shown here: https://github.com/daschuer/vcpkg/runs/4363816948?check_suite_focus=true

- #### What does your PR fix?  
  Fixes ignoring VCPKG_OSX_SYSROOT  and VCPKG_OSX_ARCHITECTURES when linking meson projects. 

This also removes a deprectaion warning and reformats io2d a ffmpeg vcpkg.json to allow updating the version database 

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
No change

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  `Yes`

daschuer committed:
[meson] add -arch and -isysroot to linker flags in case of osx crosscompile

daschuer committed:
[get_cmake_vars] add -isysroot flag independednt of crosscompiling

Even without crosscompiling it is possible to specify a VCPKG_OSX_SYSROOT or
a SDKROOT this should be popagated to autotools and meson builds as well.

daschuer committed:
[meson] avoid DEPRECATION warning about flags in the [properties] section

JonLiu1993 commented:
@daschuer ,Thanks for your pr, this is error log:
https://dev.azure.com/vcpkg/public/_build/results?buildId=63911&view=artifacts&pathAsName=false&type=publishedArtifacts

JackBoosY commented:
cc @Neumann-A for simply review this PR.

Neumann-A commented on ports/fontconfig/portfile.cmake:
```diff
@@ -67,7 +67,7 @@ file(INSTALL "${SOURCE_PATH}/COPYING" DESTINATION "${CURRENT_PACKAGES_DIR}/share
 
 # Build the fontconfig cache
-if(NOT VCPKG_TARGET_IS_WINDOWS)
+if(NOT VCPKG_TARGET_IS_WINDOWS AND TARGET_TRIPLET STREQUAL HOST_TRIPLET)
```
```suggestion
if(NOT VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_CROSSCOMPILING)
```

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -152,6 +152,17 @@ function(z_vcpkg_meson_generate_flags_properties_string out_var config_type)
         set(linker_flags "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${config_type}}")
     endif()
     z_vcpkg_meson_convert_compiler_flags_to_list(linker_flags "${linker_flags}")
+    if(VCPKG_TARGET_IS_OSX)
+        # macOS - append arch and isysroot if cross-compiling
+        if(NOT "${VCPKG_OSX_ARCHITECTURES}" STREQUAL "${VCPKG_DETECTED_CMAKE_HOST_SYSTEM_PROCESSOR}")
+            foreach(arch IN LISTS VCPKG_OSX_ARCHITECTURES)
+                vcpkg_list(APPEND linker_flags -arch "${arch}")
+            endforeach()
+        endif()
+        if(VCPKG_OSX_SYSROOT)
+            vcpkg_list(APPEND linker_flags -isysroot "${VCPKG_OSX_SYSROOT}")
+        endif()    
+    endif()
```
the flags should come from CMake instead of being manually added here.

daschuer replied:
The linker flag is already obtained form cmake, see line 152. Unfortunately it does not contain the -sysroot and the -arch parameters. AFAIK, the autotool ports are already working. That is why I have decided for this solution with a limited scope.

daschuer replied:
I have just double checked the port flatbuffers here
https://github.com/daschuer/vcpkg/actions/runs/1531891431

It already uses -arch and -isysroot as desired. My guess is that the autotools are smart enough to take them form the cxxflags when linking: 
https://github.com/daschuer/vcpkg/actions/runs/1531891431   

```
37/38] : && /Applications/Xcode_12.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -fPIC -mmacosx-version-min=11.0 -std=c++11 -stdlib=libc++ -Wall -pedantic -Werror -Wextra -Wno-unused-parameter -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -mmacosx-version-min=11.0 -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/flatc.dir/src/idl_parser.cpp.o CMakeFiles/flatc.dir/src/idl_gen_text.cpp.o CMakeFiles/flatc.dir/src/reflection.cpp.o CMakeFiles/flatc.dir/src/util.cpp.o CMakeFiles/flatc.dir/src/idl_gen_cpp.cpp.o CMakeFiles/flatc.dir/src/idl_gen_csharp.cpp.o CMakeFiles/flatc.dir/src/idl_gen_dart.cpp.o CMakeFiles/flatc.dir/src/idl_gen_kotlin.cpp.o CMakeFiles/flatc.dir/src/idl_gen_go.cpp.o CMakeFiles/flatc.dir/src/idl_gen_java.cpp.o CMakeFiles/flatc.dir/src/idl_gen_ts.cpp.o CMakeFiles/flatc.dir/src/idl_gen_php.cpp.o CMakeFiles/flatc.dir/src/idl_gen_python.cpp.o CMakeFiles/flatc.dir/src/idl_gen_lobster.cpp.o CMakeFiles/flatc.dir/src/idl_gen_lua.cpp.o CMakeFiles/flatc.dir/src/idl_gen_rust.cpp.o CMakeFiles/flatc.dir/src/idl_gen_fbs.cpp.o CMakeFiles/flatc.dir/src/idl_gen_grpc.cpp.o CMakeFiles/flatc.dir/src/idl_gen_json_schema.cpp.o CMakeFiles/flatc.dir/src/idl_gen_swift.cpp.o CMakeFiles/flatc.dir/src/flatc.cpp.o CMakeFiles/flatc.dir/src/flatc_main.cpp.o CMakeFiles/flatc.dir/src/code_generators.cpp.o CMakeFiles/flatc.dir/grpc/src/compiler/cpp_generator.cc.o CMakeFiles/flatc.dir/grpc/src/compiler/go_generator.cc.o CMakeFiles/flatc.dir/grpc/src/compiler/java_generator.cc.o CMakeFiles/flatc.dir/grpc/src/compiler/python_generator.cc.o CMakeFiles/flatc.dir/grpc/src/compiler/swift_generator.cc.o CMakeFiles/flatc.dir/grpc/src/compiler/ts_generator.cc.o -o flatc   && :
```

Neumann-A replied:
> My guess is that the autotools are smart enough

no its not, as i recently discovered.

daschuer replied:
Do you have a build log that verifies it? 
I think my build log linkec above, verifies that it is already solved.

Neumann-A replied:
See discord:
https://discord.com/channels/400588936151433218/687365466422902841/928780648884355122
vs
https://discord.com/channels/400588936151433218/687365466422902841/928781172530618448

the point is that get-cmake-vars adds it in some cases depending on the shell you are running vcpkg.

daschuer replied:
I cannot open the posted links even though I have created a discord account. Can you copy the plain text? Which autotools port id affected?

daschuer replied:
Which auto tool port is affected? If have tested some and all is working fine.

Neumann-A replied:
the discord discussion is about libpq but only if the `CMAKE_C_COMPILER` is correctly set as `CC`. currently it defaults to gcc instead of correctly using clang on osx. (see #22254)

daschuer replied:
Unless you do not install the real GNU complier from homebrew or such you have only clang. 
Apple provides a gcc and g++ binaries that are only wrappers forwarding the command line to clang.

daschuer replied:
https://www.quora.com/Why-does-the-gcc-command-on-MacOS-execute-clang

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -192,6 +201,8 @@ function(z_vcpkg_meson_generate_native_file_config config_type) #https://mesonbu
     string(APPEND native_file "VCPKG_CRT_LINKAGE = '${VCPKG_CRT_LINKAGE}'\n")
 
     string(APPEND native_file "[built-in options]\n")
+    z_vcpkg_meson_generate_flags_properties_string(native_properties "${config_type}")
+    string(APPEND native_file "${native_properties}")
```
This is probably to remove the warning? The question here is if it works properly now. Last time I checked it did not properly adjust the flags.

daschuer replied:
my meson-cross-arm-linux-debug.log looks like that:
```
[properties]
[built-in options]
c_args = ['-fPIC', '-g', '-I/home/daniel/workspace/vcpkg/installed/arm-linux/include']
cpp_args = ['-fPIC', '-g', '-I/home/daniel/workspace/vcpkg/installed/arm-linux/include']
c_link_args = ['-L/home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib']
cpp_link_args = ['-L/home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib']
```
The output is: 

[25/81] /usr/bin/arm-linux-gnueabihf-gcc -Itest/test_d1f48f11.p -Itest -I../src/2.13.94-6fd83e1cbe.clean/test -I. -I../src/2.13.94-6fd83e1cbe.clean -I/home/daniel/workspace/vcpkg/installed/arm-linux/include -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -g -DHAVE_CONFIG_H -fPIC -g -MD -MQ test/test_d1f48f11.p/test-d1f48f11.c.o -MF test/test_d1f48f11.p/test-d1f48f11.c.o.d -o test/test_d1f48f11.p/test-d1f48f11.c.o -c ../src/2.13.94-6fd83e1cbe.clean/test/test-d1f48f11.c

[59/81] /usr/bin/arm-linux-gnueabihf-gcc  -o fc-conflist/fc-conflist fc-conflist/fc-conflist.p/fc-conflist.c.o -L/home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group src/libfontconfig.a /home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib/pkgconfig/../../lib/libfreetyped.a /home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib/pkgconfig/../../lib/libbz2d.a /home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib/pkgconfig/../../lib/libpng16d.a /home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib/pkgconfig/../../lib/libz.a -lm /home/daniel/workspace/vcpkg/installed/arm-linux/debug/lib/pkgconfig/../../lib/libexpat.a -Wl,--end-group -pthread

So It looks like it is working

Here you find a MacOs and Windows run: 
https://github.com/daschuer/vcpkg/actions/runs/1525303129

daschuer replied:
But I can revert it if you think this change is risky.

ras0219-msft replied:
From the docs[1]:
>Before 0.56.0, <lang>_args and <lang>_link_args must be put in the properties section instead, else they will be ignored.

also

>Changed in 0.56.0 putting <lang>_args and <lang>_link_args in the properties section has been deprecated, and should be put in the built-in options section.

so this LGTM.

[1]: https://mesonbuild.com/Machine-files.html#meson-builtin-options

Neumann-A commented on scripts/get_cmake_vars/CMakeLists.txt:
```diff
@@ -101,8 +101,8 @@ macro(_vcpkg_adjust_flags flag_var)
                 foreach(arch IN LISTS CMAKE_OSX_ARCHITECTURES)
                     string(APPEND ${flag_var} " -arch ${arch}")
                 endforeach()
-                string(APPEND ${flag_var} " -isysroot ${CMAKE_OSX_SYSROOT}")
             endif()
+            string(APPEND ${flag_var} " -isysroot ${CMAKE_OSX_SYSROOT}")
```
this looks like cmake should already be adding the flags (answer to https://github.com/microsoft/vcpkg/pull/21772/files#r760911069 ?)

daschuer replied:
Unfortunately not.

JackBoosY commented:
Can you please resolve the file confict?

daschuer commented:
All green :-)

JackBoosY commented:
Ping @Neumann-A for reply. I'm not good at meson and your are the authority here 😃

BillyONeal commented:
> Ping @Neumann-A for reply. I'm not good at meson and your are the authority here 😃

I got their attention over Discord chat and they said we should test on a Mac and figure out why meson is getting a different answer than autotools here. Unfortunately I know next to nothing about Mac, autotools, or meson myself 😅. I asked for some help, we'll see...

daschuer commented:
> why meson is getting a different answer than autotools here:

Autotools already discovers the link flags iself see: https://github.com/microsoft/vcpkg/pull/21772#discussion_r761417148

I don't know exactly how it works, but consequently we need to add the flags for meson only to make fontconfig build. 
flatbuffers builds in any case, independent from this PR.

BillyONeal commented:
If I understand correctly @Neumann-A your comment is about autotools not being correctly set up to use all of vcpkg's options (since it is detecting some stuff itself rather than doing what we said) rather than a problem with meson proper?

JackBoosY commented:
Ping @Neumann-A for reply.

Neumann-A commented:
SDKROOT and MACOSX_DEPLOYMENT_TARGET env variables are set by vcpkg in the scripts. I don't get why `isysroot` is needed at all since the compiler should take `SDKROOT` . Maybe the problem is that `vcpkg_restore_env_variables(VARS SDKROOT MACOSX_DEPLOYMENT_TARGET)` is called at the end of configure?

daschuer commented:
What do you propose as a solution?

daschuer commented:
We can't use the enviroent variables when crosscomiling. Meson builds target libs and host tools in a single run both may require a different sysroot. So I think it is a good solution to pass the parameter explicit. CMake does it as well. Having them in the complier arguments is also helpfull when reading through the logs.

Neumann-A commented:
> Meson builds target libs and host tools in a single run 

nope. not in vcpkg. vcpkg deactivates host builds if vcpkg is crosscompiling (native.file -> none.txt). As such setting the environment is fine. 

> So I think it is a good solution to pass the parameter explicit.

but not in `vcpkg_configure_meson`. It is simply the wrong location. 

> What do you propose as a solution?

try removing 
https://github.com/microsoft/vcpkg/blob/3c3cce612bea089a18f1693bcf153476a7e0ae3a/scripts/cmake/vcpkg_configure_meson.cmake#L516

daschuer commented:
It took me quite a bit to understand the issue. I did some experiments in this branch using GitHub workflows https://github.com/daschuer/vcpkg/tree/arm64-osx-min11.0_2 because I have no access to a macOS device. 
Now I think I see clear. 

Adding the `-isysroot` flag to c_link_args and cpp_link_args was not the solution for my original issue. it was the missing arch parameter. I have discovered it as the solution for my issue by comparing the linker call generated by cmake with the one generated by meson. The different was the missing `-arch` and the missing `-isysroot` flag. In addition the meson file was in an inconsistent state, because the `-isyroot` flag was only set in the compiler flags, but you need normally both. 

```
[properties]
[built-in options]
c_args = ['-fPIC', '-mmacosx-version-min=11.0', '-arch', 'arm64', '-isysroot', '/Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk', '-g', '-I/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/include']
cpp_args = ['-fPIC', '-mmacosx-version-min=11.0', '-arch', 'arm64', '-isysroot', '/Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk', '-g', '-I/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/include']
c_link_args = ['-L/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib']
cpp_link_args = ['-L/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib']
```
This PR makes the meson build consistent with the autotools and cmake others by setting the -isysroot for linker and compiler. 

The proposed alternative of relying on a temporary set environment variable works as well, but is IMHO only the second best solution because it happens kind of manically without any trace in the log files. In addition it is confusing that the -isysroot is set for the complier which is redundant and should be removed for a consistent meson file. 

This flag is generated in the common code for autotools and meson and is required for autotools in the current state.

It in my test it works to rely on the -isysroot flags and remove the SDKROOT trick. 
Note, Meson is the only build system where vcpkg touches the SDKROOT.  

Here the cmake reference taken from detec_compiler:
```
 "/Applications/Xcode_12.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode_12.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -arch arm64 -platform_version macos 11.0.0 11.0 -syslibroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -o cmTC_e585d -search_paths_first -headerpad_max_install_names -v CMakeFiles/cmTC_e585d.dir/CMakeCCompilerABI.c.o -lSystem /Applications/Xcode_12.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a
```
and the autotools refrence taken from gettext:
```
/bin/sh ../libtool  --tag=CC   --mode=link gcc -fvisibility=hidden -fPIC -mmacosx-version-min=11.0 -arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -g  -liconv -Wl,-framework -Wl,CoreFoundation  -no-undefined -export-symbols-regex '^([^g]|g[^l]|gl[^w]|glw[^t]|glwt[^h]|glwth[^r]|glwthr[^e]|glwthre[^a]|glwthrea[^d]).*' -version-info 10:0:2 -rpath /Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib -L/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib  -o libintl.la  bindtextdom.lo dcgettext.lo dgettext.lo gettext.lo finddomain.lo hash-string.lo loadmsgcat.lo localealias.lo textdomain.lo l10nflist.lo explodename.lo dcigettext.lo dcngettext.lo dngettext.lo ngettext.lo plural.lo plural-exp.lo localcharset.lo threadlib.lo lock.lo relocatable.lo langprefs.lo localename.lo localename-table.lo log.lo printf.lo setlocale.lo setlocale-lock.lo setlocale_null.lo version.lo xsize.lo osdep.lo intl-compat.lo   
```

Conclusion: 
Adding -isysroot c_link_args and cpp_link_args is a reasonable solution. 

I will now add a commit to check if we can drop setting a temporary SDKROOT I am curious if it works.

daschuer committed:
[meson] don't set OSXROOT evironment varibale

Neumann-A commented on ports/fontconfig/portfile.cmake:
```diff
@@ -67,7 +67,7 @@ file(INSTALL "${SOURCE_PATH}/COPYING" DESTINATION "${CURRENT_PACKAGES_DIR}/share
 
 # Build the fontconfig cache
-if(NOT VCPKG_TARGET_IS_WINDOWS)
+if(NOT VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_CROSSCOMPILING)
```
This should be a separate PR and maybe fontconfig needs a host dependency on itself?

Neumann-A commented on scripts/cmake/vcpkg_configure_meson.cmake:
```diff
@@ -152,6 +152,17 @@ function(z_vcpkg_meson_generate_flags_properties_string out_var config_type)
         set(linker_flags "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${config_type}}")
     endif()
     z_vcpkg_meson_convert_compiler_flags_to_list(linker_flags "${linker_flags}")
+    if(VCPKG_TARGET_IS_OSX)
+        # macOS - append arch and isysroot if cross-compiling
+        if(NOT "${VCPKG_OSX_ARCHITECTURES}" STREQUAL "${VCPKG_DETECTED_CMAKE_HOST_SYSTEM_PROCESSOR}")
+            foreach(arch IN LISTS VCPKG_OSX_ARCHITECTURES)
+                vcpkg_list(APPEND linker_flags -arch "${arch}")
+            endforeach()
+        endif()
+        if(VCPKG_DETECTED_CMAKE_OSX_SYSROOT)
+            vcpkg_list(APPEND linker_flags -isysroot "${VCPKG_DETECTED_CMAKE_OSX_SYSROOT}")
+        endif()    
+    endif()
```
Add it to the linker flags on `get_cmake_vars/CMakeLists.txt` ?

daschuer replied:
This does not work, because it would also affect autotools ports

JackBoosY replied:
Ping @Neumann-A for reply.

JackBoosY replied:
> This does not work, because it would also affect autotools ports

Can you declare more?

daschuer replied:
Here is autotools linker invocation taken from gettext where you can see that the -arch and -isysroot flags are already set. I am afraid that they will appear twice, when setting adding them to the linker flags in get_cmake_vars/CMakeLists.txt 

```
/bin/sh ../libtool  --tag=CC   --mode=link gcc -fvisibility=hidden -fPIC -mmacosx-version-min=11.0 -arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -g  -liconv -Wl,-framework -Wl,CoreFoundation  -no-undefined -export-symbols-regex '^([^g]|g[^l]|gl[^w]|glw[^t]|glwt[^h]|glwth[^r]|glwthr[^e]|glwthre[^a]|glwthrea[^d]).*' -version-info 10:0:2 -rpath /Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib -L/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib  -o libintl.la  bindtextdom.lo dcgettext.lo dgettext.lo gettext.lo finddomain.lo hash-string.lo loadmsgcat.lo localealias.lo textdomain.lo l10nflist.lo explodename.lo dcigettext.lo dcngettext.lo dngettext.lo ngettext.lo plural.lo plural-exp.lo localcharset.lo threadlib.lo lock.lo relocatable.lo langprefs.lo localename.lo localename-table.lo log.lo printf.lo setlocale.lo setlocale-lock.lo setlocale_null.lo version.lo xsize.lo osdep.lo intl-compat.lo 
```

Neumann-A replied:
Have you checked if they are in the cmake.vars files and that is the reason why they are already set?

daschuer replied:
They are not in the cmake.vars 
```
set(VCPKG_DETECTED_CMAKE_C_FLAGS_DEBUG "-fPIC -mmacosx-version-min=15.0 -arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -g")
```
```
set(VCPKG_DETECTED_RAW_CMAKE_SHARED_LINKER_FLAGS " ")
```

It looks like auto tools applies the c flags also to the linker.  

I will crosscheck.

daschuer replied:
This is my test run with the current state https://github.com/daschuer/vcpkg/actions/runs/1705341500
This is the one with the linker flag patch
https://github.com/daschuer/vcpkg/actions/runs/1719306811

daschuer replied:
Now we have:
```
set(VCPKG_DETECTED_CMAKE_SHARED_LINKER_FLAGS_DEBUG "-arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk")
```
```
/bin/sh ../libtool  --tag=CC   --mode=link gcc -fvisibility=hidden -fPIC -mmacosx-version-min=15.0 -arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -g  -liconv -Wl,-framework -Wl,CoreFoundation  -no-undefined -export-symbols-regex '^([^g]|g[^l]|gl[^w]|glw[^t]|glwt[^h]|glwth[^r]|glwthr[^e]|glwthre[^a]|glwthrea[^d]).*' -version-info 10:0:2 -rpath /Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib -L/Users/runner/mixxx-vcpkg/installed/arm64-osx-min11.0/debug/lib -arch arm64 -isysroot /Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk -o libintl.la  bindtextdom.lo dcgettext.lo dgettext.lo gettext.lo finddomain.lo hash-string.lo loadmsgcat.lo localealias.lo textdomain.lo l10nflist.lo explodename.lo dcigettext.lo dcngettext.lo dngettext.lo ngettext.lo plural.lo plural-exp.lo localcharset.lo threadlib.lo lock.lo relocatable.lo langprefs.lo localename.lo localename-table.lo log.lo printf.lo setlocale.lo setlocale-lock.lo setlocale_null.lo version.lo xsize.lo osdep.lo intl-compat.lo   
```
With -arch and -isysroot applied twice. 

Conclusion, the current state of this PR is correct.

Neumann-A replied:
I don't share that conclusion.
The reason that is happening is because the common makefile has something along the line `$(CC) $(CPPFLAGS) $(CFLAGS) $(LDFLAGS)`. Which of course doubles the flags.

daschuer replied:
Which common make file do you reference? 

What do you propose as an alternative solution?

JackBoosY replied:
All the flags should be extract from cmake:
https://github.com/microsoft/vcpkg/blob/5ef52b5b75887fb150711f5effb221dd98b99e6f/scripts/cmake/vcpkg_configure_meson.cmake#L383

Therefore, the behavor of cmake, make, qmake and meson etc. should be same.

daschuer replied:
Sure, this is already the case. 


The issue here is that cmake and autotools take these flags from the c flags and pass them to the linker. Meson does not do it and require a special handlung.

I have demonstrated that the requested solution does apply the flags twice. The current state of the PR passes the right number of flags as desired. 

qmake is out of scope, because it does not use z_vcpkg_get_cmake_vars

If you have an alternative solution in mind that works as well, please propose.

daschuer committed:
[meson] don't set MACOSX_DEPLOYMENT_TARGET environment variable

cmake appends the -mmacosx-version-min flag c/cxx flags, overriding the
value set via VCPKG_C_FLAGS. By setting the environment variable, the
VCPKG_C_FLAGS value was used for meson builds. Now the same value is
taken for both.

daschuer commented:
The last commit fixes an issue that a different OSX_DEPLOYMENT_TARGET was used in case it was set via VCPKG_C_FLAGS

daschuer committed:
[meson/make] Add CMAKE_OSX_DEPLOYMENT_TARGET to VCPKG_DETECTED_CMAKE_C_FLAGS

This is done in the same way in CMake internally

daschuer commented:
It is still failing locally: 
```
info [0x55bd0389efd0] TrackCollectionManager - External collections are disabled in test mode
info [0x55bd0389efd0] TrackCollectionManager - Library scanner is disabled in test mode
info [0x55bd0389efd0] SeratoMarkers2 - Reading bpmlock state from 'Serato Markers2' tag data...
info [0x55bd0389efd0] SeratoMarkers2 - Reading track color from 'Serato Markers2' tag data...
warning [0x55bd0389efd0] Track - Exporting and overwriting metadata into file "/tmp/mixxx-test-NjCcZd/cover-test-jpg.mp3" even though the file has been modified at QDateTime(2022-01-20 11:29:18.314 UTC Qt::UTC) after the last library synchronization at QDateTime(2022-01-20 11:29:18.306 UTC Qt::UTC)
/home/sperry/workspace/2.3/src/test/libraryfilesynctest.cpp:253: Failure
Value of: !fileLastModifiedAfter.isValid() || fileLastModifiedAfter > fileLastModifiedBefore
  Actual: false
Expected: true
info [0x55bd0389efd0] TrackCollection - Disconnecting database
info [0x55bd0389efd0] GlobalTrackCache - Destroying instance
[  FAILED  ] LibraryFileWithMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncEnabled (24 ms)
[----------] 1 test from LibraryFileWithMetadataSyncStatusOutdatedTest (24 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (24 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LibraryFileWithMetadataSyncStatusOutdatedTest.saveTrackMetadataWithSyncEnabled

 1 FAILED TEST
```

daschuer commented:
What kind of responds do you expect? I have explained the pending topic here:  
https://github.com/microsoft/vcpkg/pull/21772#discussion_r788472365

JackBoosY commented:
> What kind of responds do you expect? I have explained the pending topic here: [#21772 (comment)](https://github.com/microsoft/vcpkg/pull/21772#discussion_r788472365)

I think we should talk with @Neumann-A since I have less knowledge with meson.
But all configuration tools should use the same amount/content of configuration, why does meson need extra configuration here?

daschuer commented:
> why does meson need extra configuration here?

cmake and autotools are taking -arch and -isysroot from the c flags and pass them to the linker. Meson does not do it. 
The cmake behavior is the reason they are not in the linker flags when fetching DETECTED variables form cmake. The autotools behaviour is the reason that we can't manual add them to the fetched cmake variables, because otherwise they are duplicated within autotool builds.

ras0219-msft reviewed (approved):
Overall, this LGTM with one remaining question:

> [adding the flags also to LDFLAGS results in] -arch and -isysroot applied twice.

Is this a problem? If not, I think it would be better to add them to both to avoid meson-specific changes. Or to look at it another way: to avoid autotools-specific omissions in the core helper.

Even if it is a problem, it almost seems like it would be more correct to remove them after the fact in the autotools helper since this is _specifically_ an autotools issue as @Neumann-A pointed out. However I think it would be reasonable to merge this PR as-is and make that larger shuffle later (in this case).

JackBoosY commented:
Ping @daschuer

daschuer commented:
> However I think it would be reasonable to merge this PR as-is and make that larger shuffle later (in this case).

Let's go with this. And keep the scope of this PR to meson.
